compiler-rt/lib/ctx_profile/CtxInstrProfiling.h

```diff
@@ -51,5 +53,189 @@ class Arena final {
   const uint64_t Size;
 };
 
+/// The contextual profile is a directed tree where each node has one parent. A
+/// node (ContextNode) corresponds to a function activation. The root of the
+/// tree is at a function that was marked as entrypoint to the compiler. A node
+/// stores counter values for edges and a vector of subcontexts. These are the
+/// contexts of callees. The index in the subcontext vector corresponds to the
+/// index of the callsite (as was instrumented via llvm.instrprof.callsite). At
+/// that index we find a linked list, potentially empty, of ContextNodes. Direct
+/// calls will have 0 or 1 values in the linked list, but indirect callsites may
+/// have more.
+///
+/// The ContextNode has a fixed sized header describing it - the GUID of the
+/// function, the size of the counter and callsite vectors. It is also an
+/// (intrusive) linked list for the purposes of the indirect call case above.
+///
+/// Allocation is expected to happen on an Arena. The allocation lays out inline
+/// the counter and subcontexts vectors. The class offers APIs to correctly
+/// reference the latter.
+///
+/// The layout is as follows:
+///
+/// [[statically-declared fields][counters vector][subcontexts vector]]
```

Comment:
I think interpreting a subcontext ContextNode as part of the same block of memory immediately following counters of width uint64_t is undefined behaviour. In this case I think we get lucky because the first element of the ContextNode struct is a GUID (uint64_t) and thus the alignment of the ContextNode struct and the alignment of each individual counter matches.

I am far from an expert on undefined behaviour and my understanding is based on reading (5) in https://en.cppreference.com/w/cpp/language/reinterpret_cast and https://en.cppreference.com/w/c/language/object. Please feel free to correct me if you think otherwise.

I think the motivation is that the single chunk based layout simplifies the instrumentation GEP computations. A few questions:
1) Can we test the ContextNode runtime implementation with sanitizers to expose any alignment issues?
2) Have you considered an alternative design which does not make it easy to run into such issues?

Reply:
I don't follow. The subcontext vector is of pointers to subcontexts. Maybe I should clarify that.

Reply:
That being said - the rest isn't by accident, but we can statically check that the alignment checks out (added). Basically:
- the Arenas are 8-aligned because that's how __sanitizer::InternalAlloc returns by default, but added the explicit flag for clarity
- the memory we start bump-allocating in the Arena is 8-aligned by construction - added a static_assert for maintainability
- the choice of types for ContextNode's header is explicit (to maintain the invariant) - and added a static assert.